governance_rules.go

```diff
@@ -119,24 +124,55 @@ type TemplatedOverrideValues struct {
 	Body    []byte
 }
 
-func (g *GovernanceRules) Get(request *http.Request, entityValues []EntityRuleValues) (rules []RuleTemplate) {
+func GetMatchingRuleTemplates(config GovernanceRulesConfig, values []EntityRuleValues) (inCohorts map[string]bool, templates []RuleTemplate) {
+	inCohorts = make(map[string]bool)
+	for _, ev := range values {
+		// check all the rules for which this entity is in a cohort
+		inCohorts[ev.Rule] = true
+		// if the rule is a matching rule, add it to the list of templates
+		if rule, ok := config.EntityRules[ev.Rule]; ok && rule.ApplyTo == "matching" {
+			templates = append(templates, RuleTemplate{rule, ev.Values})
+		}
+	}
+	return
+}
+
+func GetNotMatchingRuleTemplates(rules []moesifapi.GovernanceRule, entityId string, isInCohort map[string]bool) (templates []RuleTemplate) {
+	for _, rule := range rules {
+		if rule.ApplyTo == "not_matching" && !isInCohort[rule.ID] || rule.ApplyUnidentified && entityId == "" {
+			templates = append(templates, RuleTemplate{Rule: rule})
+		}
+	}
+	return
+}
+
+func (g *GovernanceRules) Get(request *http.Request, userValues, companyValues []EntityRuleValues, userId, companyId string) (rules []RuleTemplate) {
 	config := g.Read()
 	// in a list of rules with overrides, the last override value is what will be used in the response
 	// create a slice of rules to check in reverse priority order
 	// regex rule, company rule, user rule order, i.e. user rule overrides take priority over company, etc.
-	regexToCheck := make([]RuleTemplate, len(config.Regex))
-	// copy all regex rules first
-	for i, r := range config.Regex {
-		regexToCheck[i] = RuleTemplate{Rule: r}
-	}
-	//copy all user and company entity rules
-	for _, ev := range entityValues {
-		if rule, ok := config.EntityRules[ev.Rule]; ok {
-			regexToCheck = append(regexToCheck, RuleTemplate{rule, ev.Values})
-		}
+	regexToCheck := []RuleTemplate{}
+
+	// if a user_id is matching a cohort in any rule, it will have an EntityRuleValues entry in userValues
+	// collecting the rule ids that match the company_id in isInCohort allows us to efficiently
+	// and simply check if the current entity is in any cohort for a given rule to apply non_matching rules
+	isInCohort, matching := GetMatchingRuleTemplates(config, userValues)
+	regexToCheck = append(regexToCheck, matching...)
+	regexToCheck = append(regexToCheck, GetNotMatchingRuleTemplates(config.UserRules, userId, isInCohort)...)
+
+	isInCohort, matching = GetMatchingRuleTemplates(config, companyValues)
+	regexToCheck = append(regexToCheck, matching...)
+	regexToCheck = append(regexToCheck, GetNotMatchingRuleTemplates(config.CompanyRules, companyId, isInCohort)...)
+
+	for _, r := range config.Regex {
```

Comment:
@bakennedy is this ordering correct such that user is highest pri by being overridden last? It seems to not match the comments above.

Reply:
Did I not re-write some comments?  I'll confirm

Reply:
Yep, that's the original commentary on this code path, and fixed